pkg/logging/cri_logger.go

```diff
@@ -126,13 +140,48 @@ func ReadLogs(opts *LogViewOptions, stdout, stderr io.Writer, stopChannel chan o
 					return fmt.Errorf("failed to read log file %q: %v", logPath, err)
 				}
 				if opts.Follow {
-
+					if !continueNext {
+						return nil
+					}
 					// Reset seek so that if this is an incomplete line,
 					// it will be read again.
 					if _, err := f.Seek(-int64(len(l)), io.SeekCurrent); err != nil {
 						return fmt.Errorf("failed to reset seek in log file %q: %v", logPath, err)
 					}
 
+					if watcher == nil {
+						// Initialize the watcher if it has not been initialized yet.
+						if watcher, err = fsnotify.NewWatcher(); err != nil {
+							return fmt.Errorf("failed to create fsnotify watcher: %v", err)
+						}
+						defer watcher.Close()
+						if err := watcher.Add(dir); err != nil {
+							return fmt.Errorf("failed to watch directory %q: %w", dir, err)
+						}
+						// If we just created the watcher, try again to read as we might have missed
+						// the event.
+						continue
```

Comment:
> // If we just created the watcher, try again to read as we might have missed
						// the event.

Sorry, I didn't understand the comment

Reply:
This  copy from this commit https://github.com/kubernetes/kubernetes/pull/73041/commits/2c30eee92f8a1b461848cd0ad2b2848d908a69de

Reply:
If there are one or more WRITE before `watcher.Add` and the first event after `watcher.Add` is CREATE. If we remove the `continue`, then the reader will read data from the new file, and we may lose some data.

Reply:
the watcher will detect only events after the `watcher.Add`;  the `continue` can't go back

Reply:
> the watcher will detect only events after the `watcher.Add` (it can't watch in the past)

Yes. So if there is a WRITE before watcher.Add, apparently which can't be detected by watcher.Add. And the first event after watcher.Add is Create, then func `startTail` will return `recreated` as True, and close the old log file. At last, we may lose some logs.  So, we need the `continue`, let the reader read the old file again.